File: 02. Algorithms/02. Sorting/DNF Sort/DNF_Sort.cpp
#include <bits/stdc++.h>
using namespace std;
// Function to sort the input array where the array is assumed to have values in {0, 1, 2}
void DNF(int arr[], int arr_size)
{
    int low = 0;
    int high = arr_size - 1;
    int mid = 0;
    // We have keep iterating till all the elements are sorted
    while (mid <= high)
    {
        switch (arr[mid])
        {
        // Here mid is 0.
        case 0:
            swap(arr[low++], arr[mid++]);
            break;
        // Here mid is 1.
        case 1:
            mid++;
            break;
        // Here mid is 2.
        case 2:
            swap(arr[mid], arr[high--]);
            break;
        }
    }
}
// Now, we write the function to print array arr[]
void printArray(int arr[], int arr_size)
{
    // To iterate and print every element, we follow these steps
    for (int i = 0; i < arr_size; i++)
        cout << arr[i] << " ";
}
//Main Code
int main()
{
    // int arr[] = {0, 1, 0, 1, 2, 0, 1, 2};
    // int n = sizeof(arr) / sizeof(arr[0]);
    int n;
    cout<<"Enter the size of array:"<<endl;
    cin>>n;
    int arr[n];
    cout<<"Enter the element of array:"<<endl;
    for (int i = 0; i < n; i++)
    {
        cin>>arr[i];
    }
    cout << "Array before executing the algorithm: ";
    printArray(arr, n);
    DNF(arr, n);
    cout << "\nArray after executing the DNFS algorithm: ";
    printArray(arr, n);
    return 0;
}

File: 02. Algorithms/02. Sorting/DNF Sort/DNF_Sort.java
import java.io.*;   
class DNF {   
static void JTP(int arr[], int arr_size)   
{   
int low = 0;   
int high = arr_size - 1;   
int mid = 0, temp=0; // We use temporary variable for swapping   
while (mid <= high)   
{   
switch (arr[mid])   
{  
case 0: // Here mid pointer points is at 0.  
{   
temp = arr[low];   
arr[low] = arr[mid];   
arr[mid] = temp;   
low++;   
mid++;   
break;  
}   
case 1: // Here mid pointer points is at 1.  
mid++;   
break;   
case 2: // Here mid pointer points is at 2.  
{   
temp = arr[mid];   
arr[mid] = arr[high];   
arr[high] = temp;   
high--;   
break;   }   
}   }   }   
// Now we have to call function to print array arr[]   
static void printArray(int arr[], int arr_size)   
{   
int i;   
for (i = 0; i < arr_size; i++)   
System.out.print(arr[i]+" ");   
System.out.println("");   
}   
//Now we use driver function to check for above functions  
public static void main (String[] arguments)   
{   
int arr[] = {0, 1, 0, 1, 2, 0, 1, 2};   
int arr_size = arr.length;   
System.out.println("Array before executing the DNFS algorithm : ");   
printArray(arr, arr_size);   
JTP(arr, arr_size);   
System.out.println("\nArray after executing the DNFS algorithm : ");   
printArray(arr, arr_size);   
}   }  
